File: app/src/main/java/com/company/art_and_culture/myarts/museum_fragment/MuseumViewModel.java
package com.company.art_and_culture.myarts.museum_fragment;

import android.app.Application;

import com.company.art_and_culture.myarts.MainActivity;
import com.company.art_and_culture.myarts.pojo.Art;
import com.company.art_and_culture.myarts.pojo.ArtProvider;
import com.company.art_and_culture.myarts.pojo.Maker;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.paging.PagedList;


public class MuseumViewModel extends AndroidViewModel {

    private LiveData<PagedList<Art>> artList;
    private LiveData<Boolean> isLoading;
    private LiveData<Boolean> isListEmpty;
    private LiveData<ArtProvider> artProvider;
    private LiveData<Boolean> isLiked;
    private LiveData<ArrayList<Maker>> listMakers;
    private MuseumRepository museumRepository;
    private android.content.res.Resources res;
    private Application application;

    public MuseumViewModel(@NonNull Application application) {
        super(application);

        this.application = application;

    }

    public LiveData<PagedList<Art>> getArtList(){
        return artList;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<Boolean> getIsListEmpty() {
        return isListEmpty;
    }

    public LiveData<ArtProvider> getArtProvider() {
        return artProvider;
    }

    public LiveData<ArrayList<Maker>> getListMakers() {
        return listMakers;
    }

    public LiveData<Boolean> getArtProviderLike() {
        return isLiked;
    }

    public void writeDimentionsOnServer(Art art) {
        museumRepository.writeDimentionsOnServer(art);
    }

    public boolean refresh() {
        return museumRepository.refresh();
    }

    public void setArtProviderId(String artProviderId) {

        if(museumRepository == null) {
            museumRepository = new MuseumRepository(application, artProviderId);
            artList = museumRepository.getArtList();
            isLoading = museumRepository.getIsLoading();
            isListEmpty = museumRepository.getIsListEmpty();
            artProvider = museumRepository.getArtProvider();
            listMakers = museumRepository.getListMakers();
            isLiked = museumRepository.getArtProviderLike();
        } else
            museumRepository.setArtProviderId(artProviderId);
    }

    public void setActivity(MainActivity activity) {
        museumRepository.setActivity(activity);
    }

    public boolean likeMuseum(ArtProvider museum, String userUniqueId) {
        return museumRepository.likeMuseum(museum, userUniqueId);
    }
}

File: app/src/main/java/com/company/art_and_culture/myarts/museum_fragment/MuseumViewModel.kt
package com.company.art_and_culture.myarts.museum_fragment

import androidx.lifecycle.*
import androidx.paging.*
import com.company.art_and_culture.myarts.pojo.Art
import com.company.art_and_culture.myarts.pojo.ArtProvider
import com.company.art_and_culture.myarts.pojo.Maker
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flatMapLatest

class MuseumViewModel (
    private val repository: IMuseumRepository
    ) : ViewModel() {

    var artsFlow: Flow<PagingData<Art>>

    var isLoading: LiveData<Boolean> = repository.getIsLoading()
    var isContentEmpty: LiveData<Boolean> = repository.isContentEmpty()
    var artProvider: LiveData<ArtProvider?> = repository.getArtProvider()
    var isArtProviderLiked: LiveData<Boolean?> = repository.getIsArtProviderLiked()
    var listMakers: LiveData<ArrayList<Maker>?> = repository.getListMakers()

    private val museumId = MutableLiveData("")

    init {
        artsFlow = museumId.asFlow().flatMapLatest { repository.getArtsFlow(it)  }.cachedIn(viewModelScope)
    }

    fun setMuseumId(artProviderId: String) {
        if (this.museumId.value == artProviderId) return
        this.museumId.value = artProviderId
    }

    fun refresh() {
        ArtDataInMemory.getInstance().refresh()
        this.museumId.postValue(this.museumId.value)
    }

    fun onDestroy() {
        ArtDataInMemory.getInstance().refresh()
        repository.onDestroy()
    }

    suspend fun likeMuseum(museum: ArtProvider?, userUniqueId: String?) {
        return repository.likeMuseum(museum, userUniqueId)
    }

    suspend fun writeDimensionsOnServer(art: Art?) {
        repository.writeDimensionsOnServer(art)
    }

}



@Suppress("UNCHECKED_CAST")
class ViewModelFactory (
    private val repository: IMuseumRepository): ViewModelProvider.Factory {

        override fun <T : ViewModel> create(modelClass: Class<T>): T = MuseumViewModel(repository) as T
    }


object Injection {

    private var artProviderId: String = ""
    private var userUniqueId: String = ""

    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO

    fun init (artProviderId: String?, userUniqueId: String?) {
        if (artProviderId != null) this.artProviderId = artProviderId
        if (userUniqueId != null) this.userUniqueId = userUniqueId
    }

    private val repository: IMuseumRepository by lazy { MuseumRepository(artProviderId, userUniqueId, ioDispatcher) }

    val viewModelFactory: ViewModelFactory by lazy { ViewModelFactory(repository) }
}